File: agent/generated_skills/stripe_payments_expert/agents/adk_agent.py
import os
import json
from typing import Dict, Any

import stripe
from google.adk.agents import LlmAgent, AgentRequest, AgentResponse
from google.adk.tools import Tool

class StripePaymentAgent(LlmAgent):
    def __init__(self):
        super().__init__()
        stripe.api_key = os.getenv("STRIPE_SECRET_KEY")
        if not stripe.api_key:
            raise ValueError("STRIPE_SECRET_KEY environment variable not set.")

        self.add_tool(Tool(
            name="create_stripe_customer",
            description="Creates a new customer in Stripe.",
            parameters={
                "type": "object",
                "properties": {
                    "email": {"type": "string", "description": "Customer's email address."},
                    "name": {"type": "string", "description": "Customer's full name."}
                },
                "required": ["email"]
            }
        ))
        self.add_tool(Tool(
            name="create_stripe_payment_intent",
            description="Creates a Stripe Payment Intent.",
            parameters={
                "type": "object",
                "properties": {
                    "amount": {"type": "integer", "description": "Amount in cents (e.g., 2000 for $20.00)."},
                    "currency": {"type": "string", "description": "Three-letter ISO currency code (e.g., 'usd')."},
                    "description": {"type": "string", "description": "Description for the payment intent."}
                },
                "required": ["amount", "currency"]
            }
        ))
        self.add_tool(Tool(
            name="confirm_stripe_payment_intent",
            description="Confirms an existing Stripe Payment Intent.",
            parameters={
                "type": "object",
                "properties": {
                    "payment_intent_id": {"type": "string", "description": "ID of the Payment Intent to confirm."},
                    "payment_method_id": {"type": "string", "description": "ID of the Payment Method to use."}
                },
                "required": ["payment_intent_id", "payment_method_id"]
            }
        ))
        self.add_tool(Tool(
            name="create_stripe_checkout_session",
            description="Creates a Stripe Checkout Session for a hosted payment page.",
            parameters={
                "type": "object",
                "properties": {
                    "product_name": {"type": "string", "description": "Name of the product/service."},
                    "unit_amount": {"type": "integer", "description": "Price per unit in cents."},
                    "currency": {"type": "string", "description": "Three-letter ISO currency code (e.g., 'usd')."},
                    "quantity": {"type": "integer", "description": "Number of units."},
                    "success_url": {"type": "string", "description": "URL to redirect to after successful payment."},
                    "cancel_url": {"type": "string", "description": "URL to redirect to if payment is cancelled."}
                },
                "required": ["product_name", "unit_amount", "currency", "quantity", "success_url", "cancel_url"]
            }
        ))
        self.add_tool(Tool(
            name="handle_stripe_webhook",
            description="Verifies and processes a Stripe webhook event.",
            parameters={
                "type": "object",
                "properties": {
                    "payload": {"type": "string", "description": "The raw request body of the webhook."},
                    "signature": {"type": "string", "description": "The value of the 'Stripe-Signature' header."}
                },
                "required": ["payload", "signature"]
            }
        ))

    async def call_tool(self, tool_name: str, args: Dict[str, Any]) -> AgentResponse:
        try:
            if tool_name == "create_stripe_customer":
                customer = stripe.Customer.create(email=args["email"], name=args.get("name"))
                return AgentResponse(output=json.dumps(customer.to_dict()))

            elif tool_name == "create_stripe_payment_intent":
                payment_intent = stripe.PaymentIntent.create(
                    amount=args["amount"],
                    currency=args["currency"],
                    description=args.get("description"),
                    payment_method_types=["card"]
                )
                return AgentResponse(output=json.dumps(payment_intent.to_dict()))

            elif tool_name == "confirm_stripe_payment_intent":
                payment_intent = stripe.PaymentIntent.confirm(
                    args["payment_intent_id"],
                    payment_method=args["payment_method_id"]
                )
                return AgentResponse(output=json.dumps(payment_intent.to_dict()))

            elif tool_name == "create_stripe_checkout_session":
                checkout_session = stripe.checkout.Session.create(
                    line_items=[
                        {
                            "price_data": {
                                "currency": args["currency"],
                                "product_data": {"name": args["product_name"]},
                                "unit_amount": args["unit_amount"],
                            },
                            "quantity": args["quantity"],
                        }
                    ],
                    mode="payment",
                    success_url=args["success_url"],
                    cancel_url=args["cancel_url"],
                )
                return AgentResponse(output=checkout_session.url)

            elif tool_name == "handle_stripe_webhook":
                webhook_secret = os.getenv("STRIPE_WEBHOOK_SECRET")
                if not webhook_secret:
                    raise ValueError("STRIPE_WEBHOOK_SECRET environment variable not set.")
                
                event = stripe.Webhook.construct_event(
                    payload=args["payload"],
                    sig_header=args["signature"],
                    secret=webhook_secret
                )
                return AgentResponse(output=json.dumps(event.to_dict()))

            else:
                return AgentResponse(output=f"Unknown tool: {tool_name}")

        except stripe.error.StripeError as e:
            # Handle Stripe API errors specifically
            return AgentResponse(output=f"Stripe API Error: {e}")
        except Exception as e:
            # Handle other unexpected errors
            return AgentResponse(output=f"Error: {e}")

if __name__ == "__main__":
    # Example of how to run the agent
    # In a real ADK deployment, the framework handles running the agent
    print("Starting Stripe Payment Agent...")
    # For local testing, ensure STRIPE_SECRET_KEY and STRIPE_WEBHOOK_SECRET are set
    try:
        agent = StripePaymentAgent()
        print("Stripe Payment Agent initialized successfully.")
        # You can test individual tool calls here if needed
        # E.g., agent.call_tool("create_stripe_customer", {"email": "test@example.com", "name": "Test User"})
    except ValueError as e:
        print(f"Agent initialization failed: {e}")
    except Exception as e:
        print(f"An unexpected error occurred during agent initialization: {e}")

File: agent/generated_skills/stripe_payments_expert/agents/adk_agent.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"

	adk "google.golang.org/adk/v2"
	"github.com/stripe/stripe-go/v72"
	"github.com/stripe/stripe-go/v72/checkout/session"
	"github.com/stripe/stripe-go/v72/customer"
	"github.com/stripe/stripe-go/v72/paymentintent"
	"github.com/stripe/stripe-go/v72/webhook"
)

// Agent implements the ADK agent interface.
type StripePaymentAgent struct{}

func (a *StripePaymentAgent) Act(ctx context.Context, request *adk.AgentRequest) (*adk.AgentResponse, error) {
	// Initialize Stripe with secret key from environment
	stripe.Key = os.Getenv("STRIPE_SECRET_KEY")
	if stripe.Key == "" {
		return nil, fmt.Errorf("STRIPE_SECRET_KEY environment variable not set")
	}

	toolName := request.Tool.Name
	args := request.Tool.Args

	switch toolName {
	case "create_stripe_customer":
		return a.createCustomer(ctx, args)
	case "create_stripe_payment_intent":
		return a.createPaymentIntent(ctx, args)
	case "confirm_stripe_payment_intent":
		return a.confirmPaymentIntent(ctx, args)
	case "create_stripe_checkout_session":
		return a.createCheckoutSession(ctx, args)
	case "handle_stripe_webhook":	
		return a.handleWebhook(ctx, args)
	default:
		return nil, fmt.Errorf("unknown tool: %s", toolName)
	}
}

func (a *StripePaymentAgent) createCustomer(ctx context.Context, args map[string]string) (*adk.AgentResponse, error) {
	email := args["email"]
	name := args["name"]

	params := &stripe.CustomerParams{
		Email: stripe.String(email),
		Name:  stripe.String(name),
	}

	c, err := customer.New(params)
	if err != nil {
		return nil, fmt.Errorf("failed to create customer: %w", err)
	}

	respBytes, _ := json.Marshal(c)
	return &adk.AgentResponse{Output: string(respBytes)}, nil
}

func (a *StripePaymentAgent) createPaymentIntent(ctx context.Context, args map[string]string) (*adk.AgentResponse, error) {
	amountStr := args["amount"]
	currency := args["currency"]
	description := args["description"]

	amount, err := parseAmount(amountStr)
	if err != nil {
		return nil, fmt.Errorf("invalid amount: %w", err)
	}

	params := &stripe.PaymentIntentParams{
		Amount:             stripe.Int64(amount),
		Currency:           stripe.String(currency),
		Description:        stripe.String(description),
		PaymentMethodTypes: stripe.StringSlice([]string{"card"}),
	}

	pi, err := paymentintent.New(params)
	if err != nil {
		return nil, fmt.Errorf("failed to create payment intent: %w", err)
	}

	respBytes, _ := json.Marshal(pi)
	return &adk.AgentResponse{Output: string(respBytes)}, nil
}

func (a *StripePaymentAgent) confirmPaymentIntent(ctx context.Context, args map[string]string) (*adk.AgentResponse, error) {
	piID := args["payment_intent_id"]
	pmID := args["payment_method_id"]

	params := &stripe.PaymentIntentConfirmParams{
		PaymentMethod: stripe.String(pmID),
	}

	pi, err := paymentintent.Confirm(piID, params)
	if err != nil {
		return nil, fmt.Errorf("failed to confirm payment intent: %w", err)
	}

	respBytes, _ := json.Marshal(pi)
	return &adk.AgentResponse{Output: string(respBytes)}, nil
}

func (a *StripePaymentAgent) createCheckoutSession(ctx context.Context, args map[string]string) (*adk.AgentResponse, error) {
	productName := args["product_name"]
	unitAmountStr := args["unit_amount"]
	currency := args["currency"]
	quantityStr := args["quantity"]
	successURL := args["success_url"]
	cancelURL := args["cancel_url"]

	unitAmount, err := parseAmount(unitAmountStr)
	if err != nil {
		return nil, fmt.Errorf("invalid unit_amount: %w", err)
	}
	quantity, err := parseInt(quantityStr)
	if err != nil {
		return nil, fmt.Errorf("invalid quantity: %w", err)
	}

	params := &stripe.CheckoutSessionParams{
		Mode:       stripe.String(string(stripe.CheckoutSessionModePayment)),
		SuccessURL: stripe.String(successURL),
		CancelURL:  stripe.String(cancelURL),
		LineItems: []*stripe.CheckoutSessionLineItemParams{
			{
				Quantity: stripe.Int64(quantity),
				PriceData: &stripe.CheckoutSessionLineItemPriceDataParams{
					Currency:   stripe.String(currency),
					UnitAmount: stripe.Int64(unitAmount),
					ProductData: &stripe.CheckoutSessionLineItemPriceDataProductDataParams{
						Name: stripe.String(productName),
					},
				},
			},
		},
	}

	sess, err := session.New(params)
	if err != nil {
		return nil, fmt.Errorf("failed to create checkout session: %w", err)
	}

	return &adk.AgentResponse{Output: sess.URL}, nil
}

func (a *StripePaymentAgent) handleWebhook(ctx context.Context, args map[string]string) (*adk.AgentResponse, error) {
	// In a real application, this would receive the raw request body and signature from the HTTP request.
	// For simplicity in this agent context, we assume they are passed as strings.
	payload := args["payload"]
	signature := args["signature"]
	webhookSecret := os.Getenv("STRIPE_WEBHOOK_SECRET")

	if webhookSecret == "" {
		return nil, fmt.Errorf("STRIPE_WEBHOOK_SECRET environment variable not set")
	}

	event, err := webhook.ConstructEvent([]byte(payload), signature, webhookSecret)
	if err != nil {
		return nil, fmt.Errorf("failed to verify webhook signature: %w", err)
	}

	respBytes, _ := json.Marshal(event)
	return &adk.AgentResponse{Output: string(respBytes)}, nil
}

func parseAmount(s string) (int64, error) {
	var i int64
	_, err := fmt.Sscanf(s, "%d", &i)
	return i, err
}

func parseInt(s string) (int64, error) {
	var i int64
	_, err := fmt.Sscanf(s, "%d", &i)
	return i, err
}

func main() {
	adk.Run("stripe-payment-agent", &StripePaymentAgent{})
}
